crates/backend/src/codegen.rs

```diff
@@ -783,12 +790,44 @@ impl TryToTokens for ast::Export {
         let nargs = self.function.arguments.len() as u32;
         let attrs = &self.function.rust_attrs;
 
-        let start_check = if self.start {
-            quote! { const _ASSERT: fn() = || -> #projection::Abi { loop {} }; }
-        } else {
-            quote! {}
+        let mut checks = Vec::new();
+        if self.start {
+            checks.push(quote! { const _ASSERT: fn() = || -> #projection::Abi { loop {} }; });
         };
 
+        if let Some(class) = self.rust_class.as_ref() {
+            // little helper function to make sure the check points to the
+            // location of the function causing the assert to fail
+            let mut add_check = |token_stream| {
+                checks.push(respan(token_stream, &self.rust_name));
+            };
+
+            match &self.method_kind {
+                ast::MethodKind::Constructor => {
+                    add_check(quote! {
+                        struct CheckSupportsConstructor<T: #wasm_bindgen::__rt::marker::SupportsConstructor>(T);
```

Comment:
We should avoid defining a bunch of new structs everywhere. `CheckSupportsConstructor` can be defined in `__rt` as well and reused.